packages/labs/src/controllers/async-task.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {ReactiveElement, notEqual} from 'reactive-element';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type TaskFunction = (...args: Array<any>) => any;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type Deps = Array<any>;
+export type DepsFunction = () => Deps;
+
+/**
+ * A controller that performs an asynchronous task like a fetch when its host
+ * element updates. The controller then performs an update on the host element
+ * when the task completes. The task function must be supplied and can take a
+ * list of dependencies. The `value` can be initially set and is updated with
+ * the result of the task when it's complete. The `isPending` property can be
+ * read to determine the task's state, and the `completeTask()` promise
+ * resolves when the task completes. Here's an example:
+ *
+ * class MyElement extends ReactiveElement {
+ *   url = 'example.com/api';
+ *   id = 0;
+ *   task = new AsyncTask(
+ *     this,
+ *     (url, id) =>
+ *       fetch(`${this.url}?id=${this.id}`).then(response => response.json()),
+ *     () => [this.id, this.url],
+ *     'initial value'
+ *   );
+ *
+ *   update(changedProperties) {
+ *     super.update(changedProperties);
+ *     if (this.task.isPending) {
+ *       console.log('task pending');
+ *     } else {
+ *       console.log('task value', this.task.value);
+ *     }
+ *   }
+ * }
+ */
+export class AsyncTask {
+  private _previousDeps: Deps = [];
+  private _task: TaskFunction;
+  private _dependencies: DepsFunction;
+  private _callId = 0;
+  isPending = false;
```

Comment:
When would the controller ever stay in the "initial state"? The controller would immediately move to the `isPending` state on first update... unless _all_ of the deps are initially `undefined`? That would be pretty odd I think?

Reply:
But agree we need to try/catch the await and give a state for doing something different there. And +1 to `renderTask` helper.

Reply:
Done.

Reply:
Signaling initial state from an async function is hard.

In runAsync() I had an exception called `InitialStateException` that could be thrown before the first `await` to signal it. A result return type with a status works for other conditions, but in an async function that result still comes back from a Promise, so you don't know to not be in the pending state at first. Maybe that's ok, and initial can be modelled as a special case of complete (complete because it couldn't really start).

Another option is a gate function in addition to the task function that must return true if the task can go from initial to pending.